streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/task/CheckpointProcessor.java

```diff
@@ -52,30 +62,31 @@ public class CheckpointProcessor {
 
   @Autowired private SavePointService savePointService;
 
-  public void process(Long appId, CheckPoints checkPoints) {
-    CheckPoints.Latest latest = checkPoints.getLatest();
-    if (latest == null || latest.getCompleted() == null) {
-      return;
-    }
-    CheckPoints.CheckPoint checkPoint = latest.getCompleted();
+  public void process(Long appId, @Nonnull CheckPoints checkPoints) {
+    checkPoints.getLatestCheckpoint().forEach(checkPoint -> process(appId, checkPoint));
+  }
+
+  private void process(Long appId, @Nonnull CheckPoints.CheckPoint checkPoint) {
     Application application = applicationService.getById(appId);
     CheckPointStatus status = checkPoint.getCheckPointStatus();
 
     if (CheckPointStatus.COMPLETED.equals(status)) {
-      String cacheId = appId + "_" + application.getJobId();
-      Long latestId =
-          checkPointCache.get(
-              cacheId,
-              key -> {
-                SavePoint savePoint = savePointService.getLatest(appId);
-                return Optional.ofNullable(savePoint).map(SavePoint::getChkId).orElse(null);
-              });
-
-      if (latestId == null || latestId < checkPoint.getId()) {
-        saveSavepoint(checkPoint, application);
+      String cacheId = getCacheIdForCheckpoint(appId, application.getJobId());
+      Long latestChkId = getLatestCheckpointedId(appId, cacheId);
+      if (shouldStoreAsSavepoint(checkPoint, latestChkId)) {
+        savepointedCache.put(checkPoint.getExternalPath(), DEFAULT_FLAG_BYTE);
```

Comment:
Why using the `DEFAULT_FLAG_BYTE` as the value?

Why don't using the similar logic to checkpointCache that using the cacheId as the key and checkpointId as the value?

Reply:
> Why don't using the similar logic to checkpointCache that using the cacheId as the key and checkpointId as the value?

We can't judge the savepoint if be stored out of TTL.  From the point of view, it doesn't matter what type of value is stored.   But the appId + jobId  if  shorter than savepoint path as cache key mostly. 

I‘m hesitating whether we should take savepoint as more important than checkpoint. If we treat each other equally, on the contrary, this part of logic will become simple to share the cache with checkpoint

Reply:
> We can't judge the savepoint if be stored out of TTL.

Why can't? Doesn't it have TriggerTime? If triggerTime < current time - TTL, it should be discarded, right?

> I‘m hesitating whether we should take savepoint as more important than checkpoint.

I think it's more important due to it's triggered maunally. User might want to use it.

Reply:
> Why can't? Doesn't it have TriggerTime? If triggerTime < current time - TTL, it should be discarded, right?
> 
> >

In this assumption, the same one savepoint from REST-API response perhaps store multiple times in the TTL range..